src/types/acls.rs

```diff
@@ -0,0 +1,387 @@
+                })
+            },
+        }
+        .try_build()
+    }
+
+    /// Access to the wrapped [`ListRights`] struct
+    pub fn acl(&self) -> &Acl<'_> {
+        self.borrow_acl()
+    }
+}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// Used by [`AclResponse`].
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub struct Acl<'a> {
+    /// The mailbox the ACL Entries belong to
+    pub(crate) mailbox: Cow<'a, str>,
+    /// The list of identifier/rights pairs for the mailbox
+    pub(crate) acls: Vec<AclEntry<'a>>,
+}
+
+impl<'a> Acl<'a> {
+    /// Return the mailbox the ACL entries belong to
+    pub fn mailbox(&self) -> &str {
+        &*self.mailbox
+    }
+
+    /// Returns a list of identifier/rights pairs for the mailbox
+    pub fn acls(&self) -> &[AclEntry<'_>] {
+        &*self.acls
+    }
+}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The list of identifiers and rights for the [`Acl`] response
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct AclEntry<'a> {
+    /// The user identifier the rights are for
+    pub identifier: Cow<'a, str>,
+    /// the rights for the provided identifier
+    pub rights: AclRights,
+}
+
+/// From [section 3.7 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.7).
+///
+/// This is a wrapper around a single [`ListRights`].
+///
+/// The LISTRIGHTS response from the [`Session::list_rights`] IMAP command
+#[self_referencing]
+pub struct ListRightsResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) rights: ListRights<'this>,
+}
+
+impl ListRightsResponse {
+    /// Parse the given input into a [`ListRights`] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        ListRightsResponseTryBuilder {
+            data: owned,
+            rights_builder: |input| {
+                // There should only be ONE single LISTRIGHTS response
+                parse_until_done(input, unsolicited, |response| match response {
+                    Response::ListRights(a) => Ok(MapOrNot::Map(ListRights {
+                        mailbox: a.mailbox,
+                        identifier: a.identifier,
+                        required: a.required.into(),
+                        optional: a.optional.into(),
+                    })),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })
+            },
+        }
+        .try_build()
+    }
+
+    /// Access to the wrapped [`ListRights`] struct
+    pub fn list_rights(&self) -> &ListRights<'_> {
+        self.borrow_rights()
+    }
+}
+
+/// From [section 3.7 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.7).
+///
+/// Used by [`ListRightsResponse`].
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
```

Comment:
This doesn't need to be `#[non_exhaustive]` since it has private fields. It would need `#[non_exhaustive]` if we made all those fields `pub` (and removed the accessor methods) though.